### PR TITLE
74 feature 65 research for state machine

### LINKED ISSUE #74: [Feature]: (#65) Research for state machine
### Description

based on #65 

- research in old projects/other robotic projects what can we reuse? (mostly probabilistic)
- how is the state machine structured there?
- difference between state machine and markow chain
- difficulties to transition between them later on

### Definition of Done

- a markdown containing all the most important research information, helping to decide on the structure of the state machine

### PR DESCRIPTION
**Pull request by JoKircher**
# Description

In order to create a suitable state machine, old projects and resources have been examined as well as possible software libraries which could be useful.

I also renamed some of the old documents to be complient with the naming style.

Fixes #74 

## Time invested

Josef Kircher: 8 h

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

No

## Most important changes

04_decision_making.md

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

**Review by erlbacsi — APPROVED** [review]
Very nice resaearch about the decision making in the older projects. You get a good overview on possible solutions.
I think the SMACH library sounds really interesting and is worth to spend more time with it. It would be a good starting point. I would also investigate and keep in mind the markov approach and maybe adapt it in the next sprints, if it is promising.

**Comment by erlbacsi** [comment]
Review time: 30 min

**Review by erlbacsi — CHANGES_REQUESTED** [review]
Request changes

File: doc/03_research/03_planning/04_decision_making.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+### State machines
+
+#### SMACH
+
+* Task-level architecture for creating state machines for robot behaviour.
+* Based on Python
+* Fast prototyping: Quickly create state machines
+* Complex state machines can easily be created
+* Introspection: smach_viewer provides a visual aid to follow the state machine executing its tasks
+  * smach_viewer is unmaintained and does not work with noetic
```
Comment (vogelnik) [inline doc/03_research/03_planning/04_decision_making.md:165]: Wie problematisch ist das? Können wir das zum laufen bringen oder kommt SMACH daher eigl nicht in Frage?

File: doc/03_research/03_planning/04_decision_making.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+* ROS library for robot decision-making based on Markov Decision Problems
+* written in C++
+* callback-based action interpretation allows to use other frameworks (SMACH)
+* relatively easy to implement hierarchical MDPs
+* supports synchronous and asynchronous execution
+
+Why not use markov_decision_making:
+
+* not maintained
+* only works with ROS hydro
```
Comment (vogelnik) [inline doc/03_research/03_planning/04_decision_making.md:214]: Wie problematisch ist das? Können wir das zum laufen bringen oder kommt markov_decision_making daher eigl nicht in Frage?

**Review by vogelnik — COMMENTED** [review]
Tolle Arbeit! Recherche ergibt eine tolle Zusammenfassung und guten Überblick über Möglichkeiten der Umsetzung. 

Evtl noch TOC fixen und planning intern über finalen umsetzungsansatz diskutieren

~30 min review time

File: doc/03_research/03_planning/04_decision_making.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+Finite-state machines are of two types—deterministic finite-state machines and non-deterministic finite-state machines. A deterministic finite-state machine can be constructed equivalent to any non-deterministic one.
+
+#### Advantages
+
+* easy to implement
+* we know most of the scenarios (finite state space)
+* previous groups have solutions we could adapt/extend
+
+#### Disadvantages
+
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:62]: Good point. I added it in another commit

File: doc/03_research/03_planning/04_decision_making.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+* many states necessary
+
+### Markov Chain
+
+A Markov chain or Markov process is a stochastic model describing a sequence of possible events in which the probability of each event depends only on the state attained in the previous event.
+A countably infinite sequence, in which the chain moves state at discrete time steps, gives a discrete-time Markov chain. A continuous-time process is called a continuous-time Markov chain. It is named after the Russian mathematician Andrey Markov.
+
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:69]: That is a good idea. I think I will spend more time investigating when we get to the point where we actually want to implement MDP.

File: doc/03_research/03_planning/04_decision_making.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+* real-time ROS applications
+* written in C++
+* designed to allow programmers to build robot control applications for multicomponent robots, in an intuitive and systematic manner.
+* well maintained, lots of prebuild state machines to possibly start from
+
+Why not use SMACC:
+
+* might get some time to get back into C++
+* more sophisticated library might need more time to get used to
+* awful country music in the back of tutorial videos
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:191]: Thank you!!!

File: doc/03_research/03_planning/04_decision_making.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+#### State machine
+
+* 2 state machines: one for maneuvers, one for speed control
+* Speed control more complex, when to brake seems like the most challenging task
+
+#### Take away
+
+* Some states seem to be comparable to what we are required to accomplish by the leaderboard
+* Our task might be more complex, needs additional states and transitions
+* I'm uncertain about an extra speed state, might be easier to handle that more locally by the local planner, maybe in combination with an observer element that keeps track of the surrounding by processing the information from `Perception`
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:109]: Yes, kind of. There are three states ACCEL, BRAKE, KEEP and each state tells the Acting Controller if it should accelerate, brake or keep the speed. It also senses the speed of vehicle in front, check for traffic lights, speed limits and obstacles.

This is a design decision we have to make. Which part should control the speed. That is one thing I requested a meeting for.

File: doc/03_research/03_planning/04_decision_making.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+  * [Date](#date)
+  * [Prerequisite](#prerequisite)
+  * [Decision-making algorithms](#decision-making-algorithms)
+    * [State machine](#state-machine)
+    * [Markov Chain](#markov-chain)
+    * [Decision Tree](#decision-tree)
+  * [Previous approaches](#previous-approaches)
+    * [PAF21-1](#paf21-1)
+    * [PAF21-2](#paf21-2)
+    * [PSAF1(2020)](#psaf1--2020-)
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:30]: Thanks. I did change that in a future commit :)

File: doc/03_research/03_planning/04_decision_making.md (hunk 1 above)
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:165]: Also, es gab Teams in den letzten Jahren, die Entscheidungsbäume verwendet haben und die haben diese Eigenschaft auch nicht erfüllt. Ich denke die einzige Stelle wo es eventuell problematisch werden könnte, ist die Notbremsung. Ansonsten würde ich sagen, wir werden es nur rausfinden, wenn wir anfangen das zu implementieren

File: doc/03_research/03_planning/04_decision_making.md (hunk 2 above)
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:214]: Ich konnte die Beispiele nicht zum Laufen bringen und bin auch bisher auf keine weitere Implementierung gestoßen, ich habe aber einige selbstimplementierte Lösungen in Python gefunden, die gut verständlich waren.

File: doc/03_research/03_planning/04_decision_making.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md:62]: Resolved in [80612a4](https://github.com/ll7/paf22/pull/76/commits/80612a45ec6695d399c0c8fa7ed40272f77836ae)

File: doc/03_research/03_planning/04_decision_making.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoKircher) [reply doc/03_research/03_planning/04_decision_making.md]: Resolved in [80612a4](https://github.com/ll7/paf22/pull/76/commits/80612a45ec6695d399c0c8fa7ed40272f77836ae)

**Review by erlbacsi — APPROVED** [review]
Ready to merge

**Pull request merged**